### ISSUE TITLE
`self.log` in `on_*_epoch_end` defaults to step=global step rather than step=epoch

### ISSUE THREAD

**Issue by aamster** [Lightning-AI/pytorch-lightning#17671]
### Bug description

adding a `self.log(...)` in `on_*_epoch_end` defaults to the global step number. If we are logging after the epoch ends, I would expect the default step number to be the epoch number. 

If this is intentional, I am not sure why. 

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

```python
class Model(lightning.LightningModule):
    ...

    def on_train_epoch_end(self) -> None:
        self.log_dict({
            'train_map': self.train_map.compute()['map'].item()
        })
        self.train_map.reset()

mlf_logger = MLFlowLogger(
    ...
)
trainer = Trainer(
    logger=mlf_logger,
    max_epochs=3,
    callbacks=[TQDMProgressBar()],
    accelerator="auto",
    devices=1,
    num_sanity_val_steps=0,
    log_every_n_steps=1
)
trainer.fit(model, ...)

```

Number of steps per epoch in my case is 2.
It will log `train_map` with steps 0, 3, 5 corresponding to the steps per epoch rather than 0, 1, 2 which are the epoch numbers.
```


### Error messages and logs

```
# Error messages and logs here please
```


### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>


### More info

_No response_

**Comment by jojje** [comment]
Seems it's an off-by-one sort of bug.

When logging from `on_train_epoch_end`, the metrics are written on the _following_ epoch's stats bar.

E.g. from a 3 epoch run:
```
Epoch 0: 100%|█████████████████████████| 782/782 [00:01<00:00, 468.45it/s, v_num=181]
Epoch 1: 100%|█████████████| 782/782 [00:01<00:00, 490.17it/s, v_num=181, loss=0.434]
Epoch 2: 100%|█████████████| 782/782 [00:01<00:00, 487.26it/s, v_num=181, loss=0.219]
`Trainer.fit` stopped: `max_epochs=3` reached.
Epoch 2: 100%|█████████████| 782/782 [00:01<00:00, 483.95it/s, v_num=181, loss=0.219]
```

* Loss for epoch 0 is incorrectly shown shown for epoch 1. 
* Loss for epoch 1 is incorrectly shown for epoch 2. 
* No logged loss at all is reported for epoch 0 nor epoch 2, 

So there is both _data loss_, as well as an off-by-one progress bar<->metrics pairing problem.

Code:
```
def on_train_epoch_end(self):
    loss = torch.stack(self.accum_losses).mean()
    self.log('loss', loss, on_step=False, on_epoch=True, prog_bar=True)
    self.train_stats.clear()
    print("")
```

versions:
```
pytorch-lightning==2.1.3
torch==2.1.2+cu118
```

**Comment by SkafteNicki** [comment]
The intended behavior of `self.log` is that the specified key-value pair is always logged based on the current global step number. Reason for this is consistency, to make sure that `self.log` does not behave different in different places. However, `lightning` automatically also logs a (step, epoch) curve which makes it possible to convert between step and epoch when needed. Most loggers like weights and bias, mlflow supports supports this out of the box.
Closing issue, but feel free to ping me and reopen if necessary.
